### PR TITLE
FEATURE: Move ImageUri and ImageTag Fusion prototypes to Neos.Media and add more general prototype Image

### LINKED ISSUE #2799: Move ImageUri and ImageTag Fusion prototypes to Neos.Media
### Description

tbd (see #2770)

### PR DESCRIPTION
**Pull request by gradinarufelix**
**What I did**

- Added a new prototype `Neos.Media:Image` which returns not only the `src` of the thumbnail, but also it's width and height, based on the known `ImageUriImplementation` class.
- Moved `Neos.Neos:ImageUri` and `Neos.Neos:ImageTag` to `Neos.Media` for naming consistency reasons, rebuilt them using `Neos.Media:Image` and AFX.
- `Neos.Neos:ImageUri` and `Neos.Neos:ImageTag` still exist because of compability reasons and are now only a reference to the new prototypes. `Neos.Neos:Image` also added.

**How to verify it**
Run tests and with Neos.Demo.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Remove old ImageUriImplementation.php in Neos.Neos
- [x] Rewrite documentation 
- [x] Add migration
- [x] Add `ImageUriImplementation` class in Neos.Neos and extend it from the new implementation
- [x] Deprecate old prototypes and classes

Resolves: #2799

### CODE REVIEW

**Review by mficzel — CHANGES_REQUESTED** [review]
I generally like the direction.

It would be great if the `Neos\Media\Fusion\ImageImplementation` would 
be the moved and modified `Neos\Neos\Fusion\ImageUriImplementation`. 

That would make tacking changes and also upmerging fixes on the long run much easier.

**Comment by gradinarufelix** [comment]
`ImageUriImplementation` should be deleted anyway. So you mean instead of just copying the file and removing the file afterwards I should just move the file to keep track of changes, right?

**Comment by mficzel** [comment]
Yes, that makes clear that the core logic is the same only some more data is returned. By doing so it is easier to believe that this is really non-breaking.

File: Neos.Media/Classes/Fusion/ImageImplementation.php
Comment (mficzel) [inline Neos.Media/Classes/Fusion/ImageImplementation.php]: This should still extend the AbstractFusionObject inheriting from DataStructureImplementation does not offer any advantage.

File: Neos.Neos/Resources/Private/Fusion/Prototypes/Image.fusion
Comment (mficzel) [inline Neos.Neos/Resources/Private/Fusion/Prototypes/Image.fusion]: No need to add `Neos.Neos:Image`. `Neos.Media:Image` is there and totally fine.

File: Neos.Neos/Resources/Private/Fusion/Prototypes/Image.fusion
Comment (mficzel) [inline Neos.Neos/Resources/Private/Fusion/Prototypes/Image.fusion]: I think those two prototypes should be declared as deprecated. 

It would be cool to have a code migration that replaces those. Since this is a clear 1:1 mapping that should be easy and will allow us to safely remove them in Neos 6 or 7.

File: Neos.Media/Classes/Fusion/ImageImplementation.php
Comment (mficzel) [inline Neos.Media/Classes/Fusion/ImageImplementation.php]: :array as return type of the method would be great

File: Neos.Media/Classes/Fusion/ImageImplementation.php
Comment (mficzel) [inline Neos.Media/Classes/Fusion/ImageImplementation.php]: declare(strict_types=1);

**Review by mficzel — CHANGES_REQUESTED** [review]
Some minor nitpicks already by reading. Will test once those are adjusted.

File: Neos.Media/Classes/Fusion/ImageImplementation.php
Comment (kitsunet) [reply Neos.Media/Classes/Fusion/ImageImplementation.php]: I know that is really nitpicky, but fusion is such a public thing, I fear this might be a breaking change for people extending the implementation, and while I am usually also the first to say, so what, I have the feeling that fusion implementations are probably a prime target of being extended, just because you can replace them so easily.

Comment (gradinarufelix) [reply Neos.Media/Classes/Fusion/ImageImplementation.php]: So should I leave it with that or switch back to `AbstractFusionObject`?

Comment (mficzel) [reply Neos.Media/Classes/Fusion/ImageImplementation.php]: Hmm, not sure 

we might also just add the new Prototypes to Media and simply don’t touch the old ones but deprecate then. For Neos 6 we can then remove the implementations and add a migration to the new names and then in Neos 7 finally remove the Prototypes.

@christian would that be a valid strategy

Comment (kitsunet) [reply Neos.Media/Classes/Fusion/ImageImplementation.php]: I guess the full blown b/c with migration path strategy would be empty classes extending the existing ones + empty prototypes and deprecating the old classes and prototypes. Then move actual code over in next major and delete old stuff.

Comment (gradinarufelix) [reply Neos.Media/Classes/Fusion/ImageImplementation.php]: Then I'd go for that. Means recreating the classes and extend from the new ones. So I guess I will revert the commit where I moved the class and then do that. And add the migration.

Just how can I mark a Fusion prototype deprecated?

Comment (gradinarufelix) [reply Neos.Media/Classes/Fusion/ImageImplementation.php]: But when extending the old class from the new one, should I do the following?

```
public function evaluate(): string
{
   return super::evaluate()['src'];
}
```

Because the new class returns the full array whereas the old class just returned the src string.

Comment (mficzel) [reply Neos.Media/Classes/Fusion/ImageImplementation.php]: > Just how can I mark a Fusion prototype deprecated?
Add a deprecated notice to he source and the fusion docs

**Comment by bwaidelich** [comment]
I added an issue for this in order to simplify release management and to collect higher-level details for the release notes: #2799

**Comment by davidspiola** [comment]
@mficzel do you have time for a final review?

File: Neos.Media/Classes/Fusion/ImageImplementation.php
Comment (mficzel) [inline Neos.Media/Classes/Fusion/ImageImplementation.php]: The comment is a bit misleading

File: Neos.Media/Classes/Fusion/ImageImplementation.php
Comment (mficzel) [reply Neos.Media/Classes/Fusion/ImageImplementation.php]: I am fine with the inheritence now but it would be good to use AbstractFusionObject as base and not DataStructureImplementation. The DataStructure implements a lot of stuff we do not want / need here.

File: Neos.Media/Classes/Fusion/ImageImplementation.php
Comment (mficzel) [inline Neos.Media/Classes/Fusion/ImageImplementation.php]: This should at least be nullable as getFormat

**Review by mficzel — CHANGES_REQUESTED** [review]
I like this change nut it requires some more work. I run into quite some problems during testing which are mostly caused by the return types of the get* methods in the ImageImplementation. Other more nitpicky things are also still not 100% there.

I suggest to push this soon after 5.1 is released to get this in for 5.2 with little more time to mature.

File: Neos.Media/Classes/Fusion/ImageImplementation.php
Comment (mficzel) [inline Neos.Media/Classes/Fusion/ImageImplementation.php]: Should be nullable or habe not return type

File: Neos.Media/Classes/Fusion/ImageImplementation.php
Comment (mficzel) [inline Neos.Media/Classes/Fusion/ImageImplementation.php]: Should be nullable or habe not return type

**Comment by gradinarufelix** [comment]
@mficzel @kitsunet Can we still include this in Neos 5.2 if I solve the open change requests?

**Comment by mficzel** [comment]
@gradinarufelix no promise, testing this always takes a while and technically 5.2 is already in feature freeze. Only release managers can still allow to merge features.

**Comment by jonnitto** [comment]
@gradinarufelix / @mficzel Any news on this? Today is feature-freeze for Neos 6.0

**Comment by mficzel** [comment]
@jonnitto Nothing new from my side, since this is planned non breaking anyways there is no need to push hard for 6.0

**Comment by albe** [comment]
Needs a rebase and the typehint adjustments, but otherwise anything preventing this from getting merged? @mficzel

**Comment by mficzel** [comment]
Still needs a bit of work as i mentioned above. @gradinarufelix can you wok on it those days. I would be willing to review quickly.

**Pull request closed without merging**
